packages/world/src/systems/libraries/MoveLib.sol

```diff
@@ -85,165 +145,102 @@ library MoveLib {
   /**
    * Calculate total energy cost and final path coordinate
    */
-  function _computePathResult(Vec3 currentBaseCoord, Vec3[] memory newBaseCoords, uint128 currentEnergy)
+  function _computePathResult(Vec3 current, Vec3[] memory newBaseCoords, uint128 currentEnergy)
     internal
     view
-    returns (Vec3, uint128, uint16, bool)
+    returns (Vec3, uint128)
   {
-    uint128 totalCost = 0;
-    uint16 numJumps = 0;
-    uint16 numGlides = 0;
-    uint16 currentFallHeight = 0;
-    bool gravityApplies = false;
-
-    Vec3 oldBaseCoord = currentBaseCoord;
-    for (uint256 i = 0; i < newBaseCoords.length; i++) {
-      Vec3 nextBaseCoord = newBaseCoords[i];
-      _requireValidMove(oldBaseCoord, nextBaseCoord);
-
-      uint128 stepCost = MOVE_ENERGY_COST;
-
-      gravityApplies = _gravityApplies(nextBaseCoord);
-      if (gravityApplies) {
-        if (nextBaseCoord.y() > oldBaseCoord.y()) {
-          numJumps++;
-          require(numJumps <= MAX_PLAYER_JUMPS, "Cannot jump more than 3 blocks");
-        } else if (nextBaseCoord.y() < oldBaseCoord.y()) {
-          currentFallHeight++;
-          if (currentFallHeight >= PLAYER_FALL_DAMAGE_THRESHOLD) {
-            stepCost = PLAYER_FALL_ENERGY_COST;
-          }
-          numGlides = 0;
-        } else {
-          numGlides++;
-          require(numGlides <= MAX_PLAYER_GLIDES, "Cannot glide more than 10 blocks");
-        }
-      } else {
-        numJumps = 0;
-        numGlides = 0;
-        currentFallHeight = 0;
-      }
-
-      totalCost += stepCost;
+    uint128 cost = 0;
+    uint16 jumps = 0;
+    uint16 glides = 0;
+    uint16 fallHeight = 0;
 
-      oldBaseCoord = nextBaseCoord;
+    bool gravityApplies = false;
+    uint128 fallDamage = 0;
 
-      if (totalCost >= currentEnergy) {
-        break;
-      }
-    }
+    for (uint256 i = 0; i < newBaseCoords.length && cost < currentEnergy; i++) {
+      Vec3 next = newBaseCoords[i];
+      _requireValidMove(current, next);
 
-    return (oldBaseCoord, totalCost, currentFallHeight, gravityApplies);
-  }
+      gravityApplies = _gravityApplies(next);
 
-  function moveWithoutGravity(EntityId player, Vec3 playerCoord, Vec3[] memory newBaseCoords) public {
-    Vec3[] memory playerCoords = ObjectTypes.Player.getRelativeCoords(playerCoord);
-    EntityId[] memory players = _getPlayers(player, playerCoords);
+      int32 dy = next.y() - current.y();
 
-    // Remove the current player from the grid
-    for (uint256 i = 0; i < playerCoords.length; i++) {
-      ReverseMovablePosition._deleteRecord(playerCoords[i]);
-    }
+      if (dy < 0) {
```

Comment:
hmm without the check on `gravityApplies` dy < 0 doesnt imply a fall right

eg you're moving down a staircase
current: x, y, z
next: x - 1, y - 1, z

Reply:
Also in that case the checks for valid moves would need to allow diagonal movement

Reply:
not stairs as in blocks, I mean just stepping down one block

Reply:
Oh shit hadn't realized that, ok will fix and add some more tests

Reply:
hmm well dy < 0 without gravity still might count as a fall, for example at the end of a fall gravity doesn't apply at the last block